File: simulation/src/generation/scenarios.py
"""WRO 2026 official 36-scenario traffic sign definitions.

Each scenario defines pillar placement in the South corridor template.
Coordinates are transformed to other corridors via apply_scenario_to_section().
"""

from __future__ import annotations

from src.config.constants import TrackDimensions
from src.config.enums import Section

# WRO 2026 official 36 predefined traffic sign scenarios.
# Each entry is a list of (color, x, y) tuples for the South corridor.
# x = depth position (1.0=near, 1.5=mid, 2.0=far)
# y = width position (0.4=outer, 0.6=inner)
SCENARIOS: dict[int, list[tuple[str, float, float]]] = {
    # Single pillar scenarios (1–12)
    1:  [("green", 1.0, 0.6)],
    2:  [("red",   1.0, 0.6)],
    3:  [("green", 1.5, 0.6)],
    4:  [("red",   1.5, 0.6)],
    5:  [("green", 2.0, 0.6)],
    6:  [("red",   2.0, 0.6)],
    7:  [("green", 1.0, 0.4)],
    8:  [("red",   1.0, 0.4)],
    9:  [("green", 1.5, 0.4)],
    10: [("red",   1.5, 0.4)],
    11: [("green", 2.0, 0.4)],
    12: [("red",   2.0, 0.4)],
    # Double pillar scenarios (13–36)
    13: [("green", 1.0, 0.4), ("green", 2.0, 0.6)],
    14: [("green", 1.0, 0.4), ("red",   2.0, 0.6)],
    15: [("red",   1.0, 0.4), ("green", 2.0, 0.6)],
    16: [("green", 1.0, 0.4), ("red",   2.0, 0.6)],
    17: [("red",   1.0, 0.4), ("green", 2.0, 0.6)],
    18: [("red",   1.0, 0.4), ("red",   2.0, 0.6)],
    19: [("green", 1.0, 0.6), ("green", 2.0, 0.4)],
    20: [("green", 1.0, 0.6), ("red",   2.0, 0.4)],
    21: [("red",   1.0, 0.6), ("green", 2.0, 0.4)],
    22: [("green", 1.0, 0.6), ("red",   2.0, 0.4)],
    23: [("red",   1.0, 0.6), ("green", 2.0, 0.4)],
    24: [("red",   1.0, 0.6), ("red",   2.0, 0.4)],
    25: [("green", 1.0, 0.6), ("green", 2.0, 0.6)],
    26: [("green", 1.0, 0.6), ("red",   2.0, 0.6)],
    27: [("red",   1.0, 0.6), ("green", 2.0, 0.6)],
    28: [("green", 1.0, 0.6), ("red",   2.0, 0.6)],
    29: [("red",   1.0, 0.6), ("green", 2.0, 0.6)],
    30: [("red",   1.0, 0.6), ("red",   2.0, 0.6)],
    31: [("green", 1.0, 0.4), ("green", 2.0, 0.4)],
    32: [("green", 1.0, 0.4), ("red",   2.0, 0.4)],
    33: [("red",   1.0, 0.4), ("green", 2.0, 0.4)],
    34: [("green", 1.0, 0.4), ("red",   2.0, 0.4)],
    35: [("red",   1.0, 0.4), ("green", 2.0, 0.4)],
    36: [("red",   1.0, 0.4), ("red",   2.0, 0.4)],
}

VALID_SCENARIO_IDS: frozenset[int] = frozenset(range(1, 37))


def apply_scenario_to_section(
    scenario_id: int,
    section: Section,
) -> list[tuple[str, float, float]]:
    """Transform scenario pillar coordinates from the South template to any corridor.

    The SCENARIOS dict defines pillar positions in South-corridor frame.
    This function rotates/mirrors them to produce correct world coordinates
    for any of the four corridors.

    Args:
        scenario_id: WRO scenario number (1–36).
        section: Target corridor section.

    Returns:
        List of (color, world_x, world_y) tuples for the target corridor.

    Raises:
        ValueError: If scenario_id is outside 1–36.
    """
    if scenario_id not in VALID_SCENARIO_IDS:
        raise ValueError(f"scenario_id must be 1–36, got {scenario_id}")

    track_max = TrackDimensions.MAX_COORD
    transformed: list[tuple[str, float, float]] = []

    for color, x_south, y_south in SCENARIOS[scenario_id]:
        if section is Section.SOUTH:
            world_x, world_y = x_south, y_south
        elif section is Section.NORTH:
            world_x = x_south
            world_y = track_max - y_south
        elif section is Section.EAST:
            world_x = track_max - y_south
            world_y = x_south
        else:  # Section.WEST
            world_x = y_south
            world_y = x_south

        transformed.append((color, world_x, world_y))

    return transformed

File: platform/simgen/internal/generate/scenarios.go
// Package generate implements WRO 2026 scenario randomization.
package generate

import (
	"fmt"

	"voldemorbot/simgen/internal/simconfig"
)

// scenarioEntry is one pillar in the South-corridor template frame.
// x = depth (1.0 near, 1.5 mid, 2.0 far), y = width (0.4 outer, 0.6 inner).
type scenarioEntry struct {
	color string
	x, y  float64
}

// scenarios is the WRO 2026 official 36-scenario traffic-sign table.
// All entries are defined in South-corridor frame; use ApplyScenarioToSection
// to transform them to world coordinates for any corridor.
//
// Scenarios 1–12: single pillar.
// Scenarios 13–36: double pillar.
var scenarios = map[int][]scenarioEntry{
	// Single pillar — inner position (y=0.6)
	1: {{simconfig.ColorNameGreen, 1.0, 0.6}},
	2: {{simconfig.ColorNameRed, 1.0, 0.6}},
	3: {{simconfig.ColorNameGreen, 1.5, 0.6}},
	4: {{simconfig.ColorNameRed, 1.5, 0.6}},
	5: {{simconfig.ColorNameGreen, 2.0, 0.6}},
	6: {{simconfig.ColorNameRed, 2.0, 0.6}},
	// Single pillar — outer position (y=0.4)
	7:  {{simconfig.ColorNameGreen, 1.0, 0.4}},
	8:  {{simconfig.ColorNameRed, 1.0, 0.4}},
	9:  {{simconfig.ColorNameGreen, 1.5, 0.4}},
	10: {{simconfig.ColorNameRed, 1.5, 0.4}},
	11: {{simconfig.ColorNameGreen, 2.0, 0.4}},
	12: {{simconfig.ColorNameRed, 2.0, 0.4}},
	// Double pillar
	13: {{simconfig.ColorNameGreen, 1.0, 0.4}, {simconfig.ColorNameGreen, 2.0, 0.6}},
	14: {{simconfig.ColorNameGreen, 1.0, 0.4}, {simconfig.ColorNameRed, 2.0, 0.6}},
	15: {{simconfig.ColorNameRed, 1.0, 0.4}, {simconfig.ColorNameGreen, 2.0, 0.6}},
	16: {{simconfig.ColorNameGreen, 1.0, 0.4}, {simconfig.ColorNameRed, 2.0, 0.6}},
	17: {{simconfig.ColorNameRed, 1.0, 0.4}, {simconfig.ColorNameGreen, 2.0, 0.6}},
	18: {{simconfig.ColorNameRed, 1.0, 0.4}, {simconfig.ColorNameRed, 2.0, 0.6}},
	19: {{simconfig.ColorNameGreen, 1.0, 0.6}, {simconfig.ColorNameGreen, 2.0, 0.4}},
	20: {{simconfig.ColorNameGreen, 1.0, 0.6}, {simconfig.ColorNameRed, 2.0, 0.4}},
	21: {{simconfig.ColorNameRed, 1.0, 0.6}, {simconfig.ColorNameGreen, 2.0, 0.4}},
	22: {{simconfig.ColorNameGreen, 1.0, 0.6}, {simconfig.ColorNameRed, 2.0, 0.4}},
	23: {{simconfig.ColorNameRed, 1.0, 0.6}, {simconfig.ColorNameGreen, 2.0, 0.4}},
	24: {{simconfig.ColorNameRed, 1.0, 0.6}, {simconfig.ColorNameRed, 2.0, 0.4}},
	25: {{simconfig.ColorNameGreen, 1.0, 0.6}, {simconfig.ColorNameGreen, 2.0, 0.6}},
	26: {{simconfig.ColorNameGreen, 1.0, 0.6}, {simconfig.ColorNameRed, 2.0, 0.6}},
	27: {{simconfig.ColorNameRed, 1.0, 0.6}, {simconfig.ColorNameGreen, 2.0, 0.6}},
	28: {{simconfig.ColorNameGreen, 1.0, 0.6}, {simconfig.ColorNameRed, 2.0, 0.6}},
	29: {{simconfig.ColorNameRed, 1.0, 0.6}, {simconfig.ColorNameGreen, 2.0, 0.6}},
	30: {{simconfig.ColorNameRed, 1.0, 0.6}, {simconfig.ColorNameRed, 2.0, 0.6}},
	31: {{simconfig.ColorNameGreen, 1.0, 0.4}, {simconfig.ColorNameGreen, 2.0, 0.4}},
	32: {{simconfig.ColorNameGreen, 1.0, 0.4}, {simconfig.ColorNameRed, 2.0, 0.4}},
	33: {{simconfig.ColorNameRed, 1.0, 0.4}, {simconfig.ColorNameGreen, 2.0, 0.4}},
	34: {{simconfig.ColorNameGreen, 1.0, 0.4}, {simconfig.ColorNameRed, 2.0, 0.4}},
	35: {{simconfig.ColorNameRed, 1.0, 0.4}, {simconfig.ColorNameGreen, 2.0, 0.4}},
	36: {{simconfig.ColorNameRed, 1.0, 0.4}, {simconfig.ColorNameRed, 2.0, 0.4}},
}

// ScenarioPillar is a traffic sign pillar with world coordinates and color.
type ScenarioPillar struct {
	Color string
	X, Y  float64
}

// ApplyScenarioToSection transforms the South-frame scenario coordinates to
// world coordinates for the given corridor section.
//
// The SCENARIOS table defines pillar positions in the South corridor:
//   - x is the depth axis (along the corridor, 1.0–2.0 m)
//   - y is the width axis (across the corridor, 0.4 or 0.6 m from outer wall)
//
// The transform mirrors/rotates to produce the correct world (x, y) for each
// of the four corridors. scenarioID must be in [1, 36].
func ApplyScenarioToSection(scenarioID int, section simconfig.Section) ([]ScenarioPillar, error) {
	if scenarioID < simconfig.ScenarioIDMin || scenarioID > simconfig.ScenarioIDMax {
		return nil, fmt.Errorf(
			"scenario_id must be %d–%d, got %d",
			simconfig.ScenarioIDMin,
			simconfig.ScenarioIDMax,
			scenarioID,
		)
	}
	entries := scenarios[scenarioID]
	out := make([]ScenarioPillar, len(entries))

	const trackMax = simconfig.TrackMaxCoord
	for i, e := range entries {
		var wx, wy float64
		switch section {
		case simconfig.SectionSouth:
			wx, wy = e.x, e.y
		case simconfig.SectionNorth:
			wx, wy = e.x, trackMax-e.y
		case simconfig.SectionEast:
			wx, wy = trackMax-e.y, e.x
		case simconfig.SectionWest:
			wx, wy = e.y, e.x
		default:
			return nil, fmt.Errorf("unknown section %q", section)
		}
		out[i] = ScenarioPillar{Color: e.color, X: wx, Y: wy}
	}
	return out, nil
}
